zeroex/orderwatch/order_watcher.go

```diff
@@ -539,6 +587,48 @@ func (w *Watcher) Add(orderInfo *ordervalidator.AcceptedOrderInfo) error {
 	return nil
 }
 
+func (w *Watcher) trimOrdersAndFireEvents() error {
+	targetMaxOrders := int(maxOrdersTrimRatio * float64(w.maxOrders))
+	newMaxExpirationTime, removedOrders, err := w.meshDB.TrimOrdersByExpirationTime(targetMaxOrders)
+	if err != nil {
+		return err
+	}
+	if len(removedOrders) > 0 {
+		logger.WithFields(logger.Fields{
+			"numOrdersRemoved": len(removedOrders),
+			"targetMaxOrders":  targetMaxOrders,
+		}).Debug("removing orders to make space")
+	}
+	for _, removedOrder := range removedOrders {
+		// Fire a "REMOVED" event for each order that was removed.
+		orderEvent := &zeroex.OrderEvent{
+			OrderHash:                removedOrder.Hash,
+			SignedOrder:              removedOrder.SignedOrder,
+			FillableTakerAssetAmount: removedOrder.FillableTakerAssetAmount,
+			EndState:                 zeroex.ESOrderRemoved,
+		}
+		w.orderFeed.Send([]*zeroex.OrderEvent{orderEvent})
+		// TODO(albrow): remove order from in-memory state.
+	}
+	if newMaxExpirationTime.Cmp(w.maxExpirationTime) == -1 {
+		// Decrease the max expiration time to account for the fact that orders were
+		// removed.
+		logger.WithFields(logger.Fields{
+			"oldMaxExpirationTime": w.maxExpirationTime.String(),
+			"newMaxExpirationTime": newMaxExpirationTime.String(),
+		}).Debug("decreasing max expiration time")
+		w.maxExpirationTime = newMaxExpirationTime
+	}
+
+	return nil
+}
+
+// MaxExpirationTime returns the current maximum expiration time for incoming
+// orders.
+func (w *Watcher) MaxExpirationTime() *big.Int {
+	return w.maxExpirationTime
+}
+
```

Comment:
I'm fine with having this logic here but we should _also_ be checking if the order's expiration is greater than the current `maxExpirationTime` _after_ we validate that the message conforms to the order schema but _before_ we validate the order with an ETH RPC call. This should be done at the top of `core.go`'s `AddOrders()` and `message_handler.go`'s `HandleMessages()`. This will avoid nodes wasting RPC calls on orders that we already know won't get stored.

Reply:
@fabioberger we already do that in __core/validation.go__. See the `validateOrders` function.